File: packages/contracts/src/schemas/doctor.ts
import {
  RUNX_LOGICAL_SCHEMAS,
  type DeepReadonly,
  type UnknownRecord,
  generatedSchema,
  generatedSchemaAt,
  validateContractSchema,
} from "../internal.js";

export type DoctorDiagnosticSeverityContract = "error" | "warning" | "info";
export type DoctorRepairKindContract =
  | "create_file"
  | "replace_file"
  | "edit_yaml"
  | "edit_json"
  | "add_fixture"
  | "run_command"
  | "manual";
export type DoctorRepairConfidenceContract = "low" | "medium" | "high";
export type DoctorRepairRiskContract = "low" | "medium" | "high" | "sensitive";

export type DoctorRepairContract = DeepReadonly<{
  id: string;
  kind: DoctorRepairKindContract;
  confidence: DoctorRepairConfidenceContract;
  risk: DoctorRepairRiskContract;
  path?: string;
  json_pointer?: string;
  contents?: string;
  patch?: string;
  command?: string;
  requires_human_review: boolean;
}>;

export type DoctorLocationContract = DeepReadonly<{
  path: string;
  json_pointer?: string;
}>;

export type DoctorDiagnosticContract = DeepReadonly<{
  id: string;
  instance_id: string;
  severity: DoctorDiagnosticSeverityContract;
  title: string;
  message: string;
  target: UnknownRecord;
  location: DoctorLocationContract;
  evidence?: UnknownRecord;
  repairs: readonly DoctorRepairContract[];
}>;

export type DoctorSummaryContract = DeepReadonly<{
  errors: number;
  warnings: number;
  infos: number;
}>;

export type DoctorReportContract = DeepReadonly<{
  schema: typeof RUNX_LOGICAL_SCHEMAS.doctor;
  status: "success" | "failure";
  summary: DoctorSummaryContract;
  diagnostics: readonly DoctorDiagnosticContract[];
}>;

export const doctorV1Schema = generatedSchema<DoctorReportContract>("doctor.schema.json");
export const doctorDiagnosticSchema = generatedSchemaAt<DoctorDiagnosticContract>(
  doctorV1Schema,
  ["properties", "diagnostics", "items"],
  "doctor.diagnostics[]",
);
export const doctorLocationSchema = generatedSchemaAt<DoctorLocationContract>(
  doctorDiagnosticSchema,
  ["properties", "location"],
  "doctor.diagnostics[].location",
);
export const doctorRepairSchema = generatedSchemaAt<DoctorRepairContract>(
  doctorDiagnosticSchema,
  ["properties", "repairs", "items"],
  "doctor.diagnostics[].repairs[]",
);
export const doctorSummarySchema = generatedSchemaAt<DoctorSummaryContract>(
  doctorV1Schema,
  ["properties", "summary"],
  "doctor.summary",
);

export function validateDoctorReportContract(value: unknown, label = "doctor_report"): DoctorReportContract {
  return validateContractSchema(doctorV1Schema, value, label);
}

File: crates/runx-contracts/src/doctor.rs
use std::fmt;
use std::marker::PhantomData;

use serde::de::{self, Visitor};
use serde::{Deserialize, Serialize};

use crate::JsonObject;

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub enum DoctorReportSchema {
    #[serde(rename = "runx.doctor.v1")]
    V1,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum DoctorStatus {
    Success,
    Failure,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum DoctorDiagnosticSeverity {
    Error,
    Warning,
    Info,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum DoctorRepairKind {
    CreateFile,
    ReplaceFile,
    EditYaml,
    EditJson,
    AddFixture,
    RunCommand,
    Manual,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum DoctorRepairConfidence {
    Low,
    Medium,
    High,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum DoctorRepairRisk {
    Low,
    Medium,
    High,
    Sensitive,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct DoctorRepair {
    pub id: String,
    pub kind: DoctorRepairKind,
    pub confidence: DoctorRepairConfidence,
    pub risk: DoctorRepairRisk,
    #[serde(
        default,
        deserialize_with = "deserialize_optional_non_null",
        skip_serializing_if = "Option::is_none"
    )]
    pub path: Option<String>,
    #[serde(
        default,
        deserialize_with = "deserialize_optional_non_null",
        skip_serializing_if = "Option::is_none"
    )]
    pub json_pointer: Option<String>,
    #[serde(
        default,
        deserialize_with = "deserialize_optional_non_null",
        skip_serializing_if = "Option::is_none"
    )]
    pub contents: Option<String>,
    #[serde(
        default,
        deserialize_with = "deserialize_optional_non_null",
        skip_serializing_if = "Option::is_none"
    )]
    pub patch: Option<String>,
    #[serde(
        default,
        deserialize_with = "deserialize_optional_non_null",
        skip_serializing_if = "Option::is_none"
    )]
    pub command: Option<String>,
    pub requires_human_review: bool,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct DoctorLocation {
    pub path: String,
    #[serde(
        default,
        deserialize_with = "deserialize_optional_non_null",
        skip_serializing_if = "Option::is_none"
    )]
    pub json_pointer: Option<String>,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct DoctorDiagnostic {
    pub id: String,
    pub instance_id: String,
    pub severity: DoctorDiagnosticSeverity,
    pub title: String,
    pub message: String,
    pub target: JsonObject,
    pub location: DoctorLocation,
    #[serde(
        default,
        deserialize_with = "deserialize_optional_non_null",
        skip_serializing_if = "Option::is_none"
    )]
    pub evidence: Option<JsonObject>,
    pub repairs: Vec<DoctorRepair>,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct DoctorSummary {
    pub errors: u64,
    pub warnings: u64,
    pub infos: u64,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct DoctorReport {
    pub schema: DoctorReportSchema,
    pub status: DoctorStatus,
    pub summary: DoctorSummary,
    pub diagnostics: Vec<DoctorDiagnostic>,
}

fn deserialize_optional_non_null<'de, D, T>(deserializer: D) -> Result<Option<T>, D::Error>
where
    D: serde::Deserializer<'de>,
    T: Deserialize<'de>,
{
    struct OptionalNonNull<T>(PhantomData<T>);

    impl<'de, T> Visitor<'de> for OptionalNonNull<T>
    where
        T: Deserialize<'de>,
    {
        type Value = Option<T>;

        fn expecting(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
            formatter.write_str("an omitted field or a non-null value")
        }

        fn visit_none<E>(self) -> Result<Self::Value, E>
        where
            E: de::Error,
        {
            Err(E::custom(
                "optional doctor fields must be omitted instead of null",
            ))
        }

        fn visit_some<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
        where
            D: serde::Deserializer<'de>,
        {
            T::deserialize(deserializer).map(Some)
        }

        fn visit_unit<E>(self) -> Result<Self::Value, E>
        where
            E: de::Error,
        {
            Err(E::custom(
                "optional doctor fields must be omitted instead of null",
            ))
        }
    }

    deserializer.deserialize_option(OptionalNonNull(PhantomData))
}
